rust/opendp-ffi/build/main.rs

```diff
@@ -0,0 +1,161 @@
+use std::env;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+
+use indexmap::map::IndexMap;
+use serde::Deserialize;
+
+#[cfg(feature="python")]
+pub mod python;
+
+// a module contains functions by name
+type Module = IndexMap<String, Function>;
+
+// metadata for each function in a module
+#[derive(Deserialize, Debug)]
+pub struct Function {
+    #[serde(default)]
+    args: Vec<Argument>,
+    // metadata for return type
+    #[serde(default)]
+    ret: Argument,
+    // metadata for constructing new types based on existing types or introspection
+    #[serde(default)]
+    derived_types: Vec<Argument>,
+    // plaintext description of the function used to generate documentation
+    description: Option<String>,
+    // URL pointing to the location of the DP proof for the function
+    proof: Option<String>
+}
+
+// Metadata for function arguments, derived types and returns.
+#[derive(Deserialize, Debug, Default, Clone)]
+pub struct Argument {
+    // argument name. Optional for return types
+    name: Option<String>,
+    // c type to translate to/from for FFI. Optional for derived types
+    c_type: Option<String>,
+    // RuntimeType expressed in terms of rust types with generics.
+    // Includes various RuntimeType constructors
+    rust_type: Option<RuntimeType>,
+    // type hint- a more abstract type that all potential arguments inherit from
+    hint: Option<String>,
+    // plaintext description of the argument used to generate documentation
+    description: Option<String>,
+    // default value for the argument
+    default: Option<String>,
+    // set to true if the argument represents a type
+    #[serde(default)]
+    is_type: bool,
+    // most functions convert c_to_py or py_to_c. Set to true to leave the value as-is
+    // an example usage is _slice_as_object,
+    //  to prevent the returned AnyObject from getting converted back to python
+    #[serde(default)]
+    do_not_convert: bool
+}
+
+#[allow(dead_code)]
+impl Argument {
+    fn name(&self) -> String {
+        self.name.clone().expect("unknown name when parsing argument")
+    }
+    fn c_type(&self) -> String {
+        self.c_type.clone().expect("unknown c_type when parsing argument")
+    }
+    fn c_type_origin(&self) -> String {
+        self.c_type().split("<").next().unwrap().to_string()
+    }
+}
+
+// RuntimeType contains the metadata to generate code that evaluates to a rust type name
+#[derive(Deserialize, Debug, PartialEq, Clone)]
+#[serde(untagged)]
+enum RuntimeType {
+    // reference an existing RuntimeType
+    Name(String),
+    // get the ith subtype of an existing RuntimeType
+    Lower { root: Box<RuntimeType>, index: i32 },
+    // build a higher level RuntimeType
+    Raise { origin: String, args: Vec<Box<RuntimeType>> },
+    // construct the RuntimeType via function call
+    Function { function: String, params: Vec<String> },
```

Comment:
Non-standard input domains and input distances were giving me a lot of grief. Also, the generated code didn't give the flexibility to pass another type arg in to set the type of the input. Which is fine, because you shouldn't be able to pass a type arg in anyways, because it can only be one thing- the type defined in the measurement/transformation. Anyways, if you set a non-standard input type (one that doesn't happen to be the inferred type), this ensures things will still work.